### PR TITLE
[dynamo] fix dynamo + DTensor to work with 2d

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108329

pair debugged with @wconstab and we found some issue in both dynamo and
the TP's fsdp extension side. This PR fixes the dynamo + DTensor integration
so that the current graph break FSDP can work with tensor parallel by moving
the torch.compile after FSDP wrapping.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: test/distributed/_tensor/test_dtensor_compile.py [hunk 1]
```diff
@@ -161,22 +161,18 @@ def test_2d_fsdp_tp_compile(self):
...
-            compiled_tp,
+        fsdp_2d = FSDP(
+            tp_model2,
             process_group=fsdp_pg,
             device_id=self.rank,
             use_orig_params=True,
         )
 
+        # TODO: once aot autograd support is ready we can just use default backend
+        compiled_2d = torch.compile(fsdp_2d, backend="eager")
```
Comment (wconstab) [inline test/distributed/_tensor/test_dtensor_compile.py:175]: nit: we might also want to test inductor backend, as it does break sometimes more than eager, due to customized logic for some parts of aot-autograd

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -572,6 +572,7 @@ def get_state_from_generator():
         elif is_from_local(self.value):
             # rewrite non-primitive args/kwargs to be included in the on-the-fly prim function
             # and rewrite args to have only proxyable args, then insert call_function
+            # TODO: support cases where device_mesh + placements specified as kwargs
```
Comment (wconstab) [inline torch/_dynamo/variables/torch.py:575]: mby good to assert kwargs is empty (or that it just contains the 1 bool you expect, etc)

**Review by wconstab — APPROVED** [review]
lgtm, but definitely add the assert for kwargs, might even be easy enough to just support kwargs too

File: test/distributed/_tensor/test_dtensor_compile.py (hunk 1 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor_compile.py:175]: aot autograd is currently broken until @bdhirsh aot dispatch subclass PR landed https://github.com/pytorch/pytorch/pull/104483, will switch this test to test both aot_eager or default backend once it's there :)

File: torch/_dynamo/variables/torch.py (hunk 2 above)
Comment (wanchaol) [reply torch/_dynamo/variables/torch.py]: Actually it's not quite hard to directly support kwargs so I added that directly

**Comment by bdhirsh** [comment]
Nice!

**Pull request closed without merging**